### PR TITLE
Allow customizing `JWTProcessor` passed to `NimbusJwtDecoder`

### PR DESCRIPTION
**Pull request by jrehwaldt**
Add ability to post-process Nimbus's `JWTProcessor` before passing it to `NimbusJwtDecoder` when using the builders.

Fixes gh-8730

### CODE REVIEW

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java [hunk 1]
```diff
@@ -461,7 +476,9 @@ public PublicKeyJwtDecoderBuilder signatureAlgorithm(SignatureAlgorithm signatur
 		 * @return the configured {@link NimbusJwtDecoder}
 		 */
 		public NimbusJwtDecoder build() {
-			return new NimbusJwtDecoder(processor());
+			ConfigurableJWTProcessor<SecurityContext> processor = processor();
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java:479]: It seems like the delta would be less if you had `processor()` use the `jwtProcessorCustomizer` instead. Any reason that it needs to be inside the `build()` method?

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java [hunk 2]
```diff
@@ -437,7 +439,20 @@ public PublicKeyJwtDecoderBuilder signatureAlgorithm(SignatureAlgorithm signatur
 			return this;
 		}
 
-		JWTProcessor<SecurityContext> processor() {
+		/**
+		 * Allows to customize the {@link JWTProcessor ConfigurableJWTProcessor} before passing it to
+		 * the build {@link NimbusJwtDecoder}.
+		 *
+		 * @param jwtProcessorCustomizer the callback used to alter the processor
+		 * @return a {@link PublicKeyJwtDecoderBuilder} for further configurations
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java:447]: Please add a `@since 5.4` to any public methods that you add in this PR.

File: oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java [hunk 3]
```diff
@@ -346,6 +347,23 @@ public void decodeWhenSignatureMismatchesAlgorithmThenThrowsException() throws E
 				.isInstanceOf(BadJwtException.class);
 	}
 
+	// gh-8730
+	@Test
+	public void decodeWhenCustomizingJwtProcessor() throws Exception {
+		RSAPublicKey publicKey = TestKeys.DEFAULT_PUBLIC_KEY;
+		RSAPrivateKey privateKey = TestKeys.DEFAULT_PRIVATE_KEY;
+		JWSHeader header = new JWSHeader.Builder(JWSAlgorithm.RS256).type(new JOSEObjectType("custom")).build();
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java:355]: Is your use case that `typ` is set to `JWS`? If so, you might consider changing the test to that to prove that it'll work for your use case.

File: oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java [hunk 4]
```diff
@@ -346,6 +347,23 @@ public void decodeWhenSignatureMismatchesAlgorithmThenThrowsException() throws E
 				.isInstanceOf(BadJwtException.class);
 	}
 
+	// gh-8730
+	@Test
+	public void decodeWhenCustomizingJwtProcessor() throws Exception {
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java:352]: While I agree that this test name makes sense, you might consider naming it to something that reflects the use case you are wanting the functionality to address, e.g. `decodeWhenCustomTypHeaderThenProcesses` or something like that.

**Comment by jrehwaldt** [comment]
@jzheaux Thank you for the comments. I incorporated your advice and added the functionality to the other two builder.

I had to "swap" the logic in the test `withJwkSetUriWhenUsingCustomTypeHeaderThenRefuseOmittedType` for the following reason: `NimbusJwtDecoderTests.JWK_SET` does not match `TestKeys.DEFAULT_PRIVATE_KEY`, which makes it impossible to construct a signed JWT with custom header `typ`. If you have a better idea I am happy to improve/align this test with the other two.

File: oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java [hunk 5]
```diff
@@ -524,6 +573,30 @@ public void decodeWhenCacheIsConfiguredAndValueLoaderErrorsThenThrowsJwtExceptio
 				.hasMessageContaining("An error occurred while attempting to decode the Jwt");
 	}
 
+	// gh-8730
+	@Test
+	public void withJwkSetUriWhenUsingCustomTypeHeaderThenRefuseOmittedType() throws Exception {
```
Comment (jrehwaldt) [inline oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java:578]: I had to "swap" the logic in the test withJwkSetUriWhenUsingCustomTypeHeaderThenRefuseOmittedType for the following reason: `NimbusJwtDecoderTests.JWK_SET` does not match `TestKeys.DEFAULT_PRIVATE_KEY`, which makes it impossible to construct a signed JWT with custom header typ. If you have a better idea I am happy to improve/align this test with the other two.

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java [hunk 6]
```diff
@@ -282,6 +284,20 @@ public JwkSetUriJwtDecoderBuilder cache(Cache cache) {
 			return this;
 		}
 
+		/**
+		 * Allows to customize the {@link JWTProcessor ConfigurableJWTProcessor} before passing it to
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java:288]: Will you please reword this to match the JavaDoc for the other methods? Maybe something closer to:

```
Use the given {@link Consumer} to customize the {@link JWTProcessor ConfigurableJWTProcessor} before ...
```

File: oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java (hunk 5 above)
Comment (jzheaux) [reply oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java:578]: I think this is fine. Note that the major goal of a test like this is just to make sure that your customization is used. So Mockito would be fine in this case, too; just call `verify` at the end of the test.

Comment (jrehwaldt) [reply oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java:578]: I opted against Mockito since it doesn't guarantee the `jwtProcessor` passed to the callback is actually the one used to verify (aka passed to `NimbusJwtDecorder`). Code like `jwtProcessorCustomizer.accept(new DefaultJWTProcessor())` would verify just fine with Mockito.

**Comment by jrehwaldt** [comment]
@jzheaux Let me know if there's anything else you'd like me to improve.

File: oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java [hunk 7]
```diff
@@ -312,6 +328,8 @@ public JwkSetUriJwtDecoderBuilder cache(Cache cache) {
 			// Spring Security validates the claim set independent from Nimbus
 			jwtProcessor.setJWTClaimsSetVerifier((claims, context) -> { });
 
+			jwtProcessorCustomizer.accept(jwtProcessor);
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java:331]: For consistency with the rest of the code here, will you please prefix member variable references with `this.`?

File: oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java [hunk 8]
```diff
@@ -346,6 +347,30 @@ public void decodeWhenSignatureMismatchesAlgorithmThenThrowsException() throws E
...
+		RSAPublicKey publicKey = TestKeys.DEFAULT_PUBLIC_KEY;
+		RSAPrivateKey privateKey = TestKeys.DEFAULT_PRIVATE_KEY;
+		JWSHeader header = new JWSHeader.Builder(JWSAlgorithm.RS256).type(new JOSEObjectType("JWS")).build();
+		JWTClaimsSet claimsSet = new JWTClaimsSet.Builder()
+				.expirationTime(Date.from(Instant.now().plusSeconds(60)))
+				.build();
+		SignedJWT signedJwt = signedJwt(privateKey, header, claimsSet);
+		NimbusJwtDecoder decoder = withPublicKey(publicKey)
+				.signatureAlgorithm(SignatureAlgorithm.RS256)
+				.jwtProcessorCustomizer(p -> p.setJWSTypeVerifier((type, context) -> {}))
```
Comment (jzheaux) [inline oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoderTests.java:362]: I'd recommend that the test follow the same allowlisting pattern that Nimbus is using. For example, it'd be more secure to do:

```java
.jwtProcessorCustomizer(p -> p.setJWSTypeVerifier(
        new DefaultJOSEObjectTypeVerifier(new JOSEObjectType("JWS"))
    )
)
```

**Comment by jrehwaldt** [comment]
@jzheaux I applied the last comments and squashed the commits.

**Comment by jzheaux** [comment]
Great, @jrehwaldt, thanks for your contribution! This is now merged into `master` via d31fff11b3bb91e484228f29af469a9a0661dd32.

Note that I did a small polish on the commit so that the title was less than 50 characters and the final message was `Closes gh-xxxx` instead of `Fixes gh-xxxx`.

I hope to see more contributions from you in the future!
